graf2d/gpadv7/v7/inc/ROOT/RColor.hxx

```diff
@@ -16,257 +16,61 @@
-      kAuto ///< The color will be set upon drawing the canvas choosing a `RPalette` color, see `RColor(Auto_t)`
-   };
-
-private:
-   // TODO: use a `variant` here!
-   /// The "R" in RGBA (0 <= R <= 1), or the palette pos if fKind is `kPalettePos`.
-   float fRedOrPalettePos = 0.;
-
-   /// The "G" in RGBA (0 <= G <= 1). Unused if `fKind != kRGBA`.
-   float fGreen = 0.;
-
-   /// The "B" in RGBA (0 <= B <= 1). Unused if `fKind != kRGBA`.
-   float fBlue = 0.;
-
-   /// The "A" in RGBA (0 <= A <= 1). Unused if `fKind != kRGBA`. `fAlpha == 0` means so transparent it's invisible,
-   /// `fAlpha == 1` means completely opaque.
-   float fAlpha = 1.;
-
-   /// How the color is defined.
-   EKind fKind = EKind::kRGBA;
-
-   /// throw an exception if the color isn't specified as `kRGBA` or `kAuto`, the two cases where
-   /// asking for RBGA members makes sense.
-   bool AssertNotPalettePos() const;
-
-public:
-   using RGBA = std::array<float, 4>;
-
-   // Default constructor: good old solid black.
-   constexpr RColor() = default;
-
-   /// Initialize a RColor with red, green, blue and alpha component.
-   constexpr RColor(float r, float g, float b, float alpha): fRedOrPalettePos(r), fGreen(g), fBlue(b), fAlpha(alpha) {}
-
-   /// Initialize a RColor with red, green, blue and alpha component.
-   constexpr RColor(float r, float g, float b, Alpha alpha = kOpaque): RColor(r, g, b, alpha.fVal) {}
-
-   /// Initialize a RColor with red, green, blue and alpha component as an array.
-   constexpr RColor(const RGBA &rgba): RColor(rgba[0], rgba[1], rgba[2], rgba[3]) {}
-
-   /// Initialize a `RColor` with a `RPalette` ordinal. The actual color is determined from the pad's
-   /// (or rather its `RFrame`'s) `RPalette`
-   constexpr RColor(float paletteOrdinal): fRedOrPalettePos(paletteOrdinal), fKind(EKind::kPalettePos) {}
-
-   /**\class AutoTag
-    Used to signal that this color shall be automatically chosen by the drawing routines, by picking a color
-    from the `RPad`'s (or rather its `RFrame`'s) current `RPalette`.
-   */
-   class AutoTag {};
-
-   /// Constructs an automatically assigned color. Call as `RColor col(RColor::kAuto)`.
-   constexpr RColor(AutoTag): fKind(EKind::kAuto) {}
+// TODO: see also imagemagick's C++ interface for RColorOld operations!
+// https://www.imagemagick.org/api/magick++-classes.php
 
-   /// Determine whether this RColor is storing RGBA (in contrast to an ordinal of a RPalette).
-   bool IsRGBA() const { return fKind == EKind::kRGBA; }
+class RColor : public RAttrBase {
 
-   /// Determine whether this `RColor` is storing an ordinal of a RPalette (in contrast to RGBA).
-   bool IsPaletteOrdinal() const { return fKind == EKind::kPalettePos; }
+   R_ATTR_CLASS(RColor, "color_", AddString("rgb","0,0,0").AddDouble("a",1.));
 
-   /// Determine whether this `RColor` will be assigned a actual color upon drawing.
-   bool IsAuto() const { return fKind == EKind::kAuto; }
```

Comment:
Before, a `RColor` could represent either:
- a set of rgb[a] values
- an index in a palette
- a flag saying "automatic coloring when drawing"
Your new `RColor` covers really only the first part. But we'll need to continue to support the last two. How would we do that?

Reply:
I propose to change RColor after PR - there are many functions which are not yet really used. Once we start using new palette in histograms drawings, we could step-by-step enable many methods again